### PR TITLE
Core Location

### PR DESCRIPTION
**Pull request by jamietsao**
Things to note:

1) Created a LocationService that is a singleton wrapper around the Core Location framework.
2) If user disabled access in the past, added a UIAlert view popup to allow them to easily navigate to Settings page to renable
3) Currently storing the latest current location in LocationService.  I'm open to changing that - depends on how we want to manage "session state" (similar to mediaArray)
4) There are still some edge cases around core location that I'd like to discuss with you guys.

### CODE REVIEW

File: Mapstagram/Info.plist [hunk 1]
```diff
@@ -2,6 +2,8 @@
 <!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
 <plist version="1.0">
 <dict>
+	<key>NSLocationWhenInUseUsageDescription</key>
+	<string>We use your location in order to find awesome Instagram photos taken around you.</string>
```
Comment (jamietsao) [inline Mapstagram/Info.plist]: We should review all of our copy as part of our polish phase.  I just put something here for now.

File: Mapstagram/Services/LocationService.swift [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+            // already authorized
+            return
+        case .NotDetermined:
+            // first time opening app so request authorization
+            locationManager.requestWhenInUseAuthorization()
+        case .Restricted, .Denied :
+            // authorization was previously denied so show alert
+            let alertController = UIAlertController(
+                title: "Location Services Disabled",
+                message: "In order to be see Instagram photos in your location, please open this app's settings and set location access to 'While Using the App'.",
```
Comment (jamietsao) [inline Mapstagram/Services/LocationService.swift]: Same comment here about copy

File: Mapstagram/Services/LocationService.swift [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+        
+        // get latest location
+        let location: CLLocation = locations.last as CLLocation
+        println("New location: \(location)")
+        self.currentLocation = location
+        self.onNewLocation(location)
+    }
+    
+    func locationManager(manager: CLLocationManager!, didFailWithError error: NSError!) {
+        println("Failed to get location: \(error)")
```
Comment (jamietsao) [inline Mapstagram/Services/LocationService.swift]: I noticed that it occasionally fails to get the location when the app starts.  Need to post on FB group to see why.

File: Mapstagram/Services/LocationService.swift [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    private var onNewLocation: ((CLLocation) -> ())!
+    private var onLocationFailure: ((NSError) -> ())!
+    
+    // TODO: not sure if this is the right place to store this (depending on how we design the common mediaArray state for all view controllers)
+    private var currentLocation: CLLocation!
+    
+    private override init() {
+        super.init()
+        
+        locationManager.desiredAccuracy = kCLLocationAccuracyBest // TODO: change accuracy?
```
Comment (kristen) [inline Mapstagram/Services/LocationService.swift]: We should probably change this. Didn't Win say that this one kills the users battery?

**Comment by kristen** [comment]
I think it looks great. I did make a comment about using kCLLocationAccuracyBest, but we can change that later, or maybe I'm remembering what Win said incorrectly. The copy looks good, we can change it later if we want.

Looks good 👍

File: Mapstagram/Services/LocationService.swift (hunk 4 above)
Comment (jamietsao) [inline Mapstagram/Services/LocationService.swift]: Yeah, I think this requires the most battery but given that we're not tracking location constantly, I figured it was ok.  It's something we can experiment with.  I left this TODO to remind me.  =)

**Pull request merged**